### PR TITLE
Add support for EvtFormatMessage and EvtCreateRenderContext

### PR DESCRIPTION
**Pull request by imron**
This commit adds previously unimplemented functions for the newer EvtLog
API - EvtFormatMessage, EvtCreateRenderContext.

It also adds EvtRenderEventValues support to the EvtRender method.

### CODE REVIEW

**Comment by imron** [comment]
Note: The error in the build check appears to be unrelated to this PR.

**Comment by EccoTheFlintstone** [comment]
really interested by this feature too, any idea when it could be implemented?

**Comment by imron** [comment]
Technically it's already implemented - you'd just need to trust the code in my pull request :-D

**Comment by EccoTheFlintstone** [comment]
I know, it was more a ping for @mhammond if he could find some time to review this PR ;)
(and it would be included in the next release this way :D )

File: win32/src/win32evtlog.i [hunk 1]
```diff
@@ -511,6 +512,113 @@ static PyObject *PyEvtOpenChannelEnum(PyObject *self, PyObject *args, PyObject *
...
+	PyObject* ret = PyList_New( 0 );
+	WCHAR* cur = strings;
+	while ( *cur ) {
+		PyObject* keyword = PyWinObject_FromWCHAR( cur );
+		PyList_Append( ret, keyword );
+		Py_DECREF( keyword );
+		cur += wcslen( cur ) + 1;
+	}
+
+	return ret;
```
Comment (mhammond) [inline win32/src/win32evtlog.i]: This function should handle failure a bit better - `PyList_New` and `PyWinObject_FromWCHAR` might return NULL - it's probably enough to early return if `ret` is null and to use `Py_XDECREF()` (although I guess even the former isn't strictly necessary as `PyList_Append` will not crash if either arg is null - but Py_DECREF will)

File: win32/src/win32evtlog.i [hunk 2]
```diff
@@ -511,6 +512,113 @@ static PyObject *PyEvtOpenChannelEnum(PyObject *self, PyObject *args, PyObject *
 }
 PyCFunction pfnPyEvtOpenChannelEnum = (PyCFunction) PyEvtOpenChannelEnum;
 
+
+//Helper function to convert a list of strings double zero terminated
```
Comment (mhammond) [inline win32/src/win32evtlog.i]: Could you please add spaces between the slash and the first letter of the comment, and remove the spaces after parens? (throughout the entire patch, not just here)

File: win32/src/win32evtlog.i [hunk 3]
```diff
@@ -511,6 +512,113 @@ static PyObject *PyEvtOpenChannelEnum(PyObject *self, PyObject *args, PyObject *
...
+		if ( flags == EvtFormatMessageKeyword ) {
+			allocated_size += 1; //+1 to double terminate the keyword list
+		}
+
+		allocated_size *= sizeof(WCHAR);
+		buf=(WCHAR *)malloc(allocated_size);
+		if (buf==NULL){
+			PyErr_NoMemory();
+			return NULL;
+			}
```
Comment (mhammond) [inline win32/src/win32evtlog.i]: this brace is misaligned

File: win32/src/win32evtlog.i [hunk 4]
```diff
@@ -511,6 +512,113 @@ static PyObject *PyEvtOpenChannelEnum(PyObject *self, PyObject *args, PyObject *
...
+
+		Py_BEGIN_ALLOW_THREADS
+		bsuccess = EvtFormatMessage(metadata, event, resourceId, 0, NULL, flags, allocated_size, buf, &returned_size);
+		Py_END_ALLOW_THREADS
+
+		if ( bsuccess ) {
+			if ( flags == EvtFormatMessageKeyword ) {
+				buf[returned_size] = L'\0';
+			}
+		}
```
Comment (mhammond) [inline win32/src/win32evtlog.i]: Please move the else to the same line as the close brace

File: win32/src/win32evtlog.i [hunk 5]
```diff
@@ -824,25 +932,119 @@ static PyObject *PyEvtSeek(PyObject *self, PyObject *args, PyObject *kwargs)
...
+
+	//bsuccess should always be false here, because we call it initially to get
+	//the size of the buffer
+	if (!bsuccess) {
+		err = GetLastError();
+		if ( err != ERROR_INSUFFICIENT_BUFFER ) {
+			PyWin_SetAPIError( "EvtRender", err );
+			goto cleanup;
+		}
+		else {
```
Comment (mhammond) [inline win32/src/win32evtlog.i]: please drop this entire else block

File: win32/src/win32evtlog.i [hunk 6]
```diff
@@ -824,25 +932,119 @@ static PyObject *PyEvtSeek(PyObject *self, PyObject *args, PyObject *kwargs)
 }
 PyCFunction pfnPyEvtSeek = (PyCFunction) PyEvtSeek;
 
-// @pyswig str|EvtRender|Formats an event into XML text
+// @pyswig o|EvtCreateRenderContext|Creates a render context
```
Comment (mhammond) [inline win32/src/win32evtlog.i]: I think the `o` here should be `<o PyEVTLOG_HANDLE>`?

File: win32/src/win32evtlog.i [hunk 7]
```diff
@@ -824,25 +932,119 @@ static PyObject *PyEvtSeek(PyObject *self, PyObject *args, PyObject *kwargs)
...
+
+	if ( !bsuccess ) {
+		err = GetLastError();
+		PyWin_SetAPIError( "EvtRender", err );
+		goto cleanup;
+	}
+
+	ret = PyList_New( prop_count );
+	for ( DWORD i = 0; i < prop_count; ++i ) {
+		PyObject* item = PyWinObject_FromEVT_VARIANT( &variants[i] );
```
Comment (mhammond) [inline win32/src/win32evtlog.i]: This loop might leave a pending python error, which should be avoided. Is there a reason to not fail entirely here? If so, we should probably `PyErr_Clear()`

File: win32/src/win32evtlog.i [hunk 8]
```diff
@@ -1010,11 +1212,116 @@ static PyObject *PyEvtUpdateBookmark(PyObject *self, PyObject *args, PyObject *k
...
+				break;
+				}
+			case EvtVarTypeEvtXml:
+				obval = PyWinObject_FromWCHAR(val->XmlValArr[i]);
+				break;
+			default:
+				PyErr_Format(PyExc_NotImplementedError, "EVT_VARIANT_TYPE %d not supported yet", val_type);
+				break;
+		    }
+
```
Comment (mhammond) [inline win32/src/win32evtlog.i]: Same comment as the loop above - and in this case you are formatting an error but not using it.

**Comment by imron** [comment]
Thanks for the review.  I'll put together an update that addresses those points in the next day or two.

**Comment by EccoTheFlintstone** [comment]
looking forward to it :D

**Comment by mhammond** [comment]
Feel free to reopen this is you get back to it.

**Comment by ofek** [comment]
@imron Any chance you could pick this back up? We're very interested in this functionality.

@mhammond Do you know of another way to parse the message from an event that works now?

**Comment by imron** [comment]
Yes, I will pick this back up at some point, it’s just behind a bunch of other higher priority things at the moment.

**Comment by ofek** [comment]
@imron Mind if I open up another PR to pick this back up (and retain your commit credit)?

**Comment by ofek** [comment]
Continued! https://github.com/mhammond/pywin32/pull/1524

**Comment by imron** [comment]
Not at all!

**Pull request closed without merging**